File: Rust/Easy/0409_longest_palindrome.rs
use std::collections::HashMap;

impl Solution {
    pub fn longest_palindrome(s: String) -> i32 {
        let mut map = HashMap::new();
        let mut count: i32 = 0;
        s.chars().for_each(|letter| {
            map.entry(letter).and_modify(|e| *e += 1).or_insert(1);
            if let Some(val) = map.get(&letter) {
                if val % 2 == 0 {
                    count += 2;
                }
            }
        });
        if s.len() as i32 - count != 0 {count + 1} else {count}
    }
}

File: python/easy/0409_longest_palindrome.py
from collections import defaultdict

class Solution:
    def longestPalindrome(self, s: str) -> int:
        freqs = defaultdict(int)
        for ch in s:
            freqs[ch] += 1
        oneTaken = False
        answer = 0
        for freq in freqs.values():
            if freq % 2 == 0:
                answer += freq
            else:
                if freq > 2:
                    answer += freq - 1
                if not oneTaken:
                    answer += 1
                    oneTaken = True
        return answer
